pkg/apis/core/validation/validation.go

```diff
@@ -4106,12 +4160,51 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		return allErrs
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		//TODO(vinaykul): Investigate if we can use PodStatus.QOSClass in kubelet and elsewhere
+		//    Once we can rely on Status persistence, drop this block. QOSClass does not change
+		//    once it is bootstrapped on podCreate. This needs to be addressed before beta.
+		//    Ref: https://github.com/kubernetes/kubernetes/pull/102884/#discussion_r663280487
+		// reject attempts to change pod qos
+		oldQoS := qos.GetPodQOS(oldPod)
+		newQoS := qos.GetPodQOS(newPod)
+		if newQoS != oldQoS {
+			allErrs = append(allErrs, field.Invalid(fldPath, newQoS, "Pod QoS is immutable"))
+		}
+	}
+
 	// handle updateable fields by munging those fields prior to deep equal comparison.
 	mungedPodSpec := *newPod.Spec.DeepCopy()
 	// munge spec.containers[*].image
 	var newContainers []core.Container
 	for ix, container := range mungedPodSpec.Containers {
 		container.Image = oldPod.Spec.Containers[ix].Image // +k8s:verify-mutation:reason=clone
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
```

Comment:
This whole block is already under that condition (see L4163)

Reply:
L4163 is before the for loop. It is checked once for the whole pod (and hopefully gone entirely once kubelet and friends can rely on bootstrapped PodStatus.QOSClass)
Within the for loop, this will look clean once we lose the featuregate.